python_modules/libraries/dagster-managed-elements/dagster_managed_elements/types.py

```diff
@@ -0,0 +1,209 @@
+            click.style(
+                f"{' ' * indent}~ {k}: {_sanitize(k, old_v)} -> {_sanitize(k, new_v)}", fg="yellow"
+            )
+            for k, old_v, new_v in self.modifications
+        ]
+
+        # Get entries for each nested diff
+        for key, nested_diff in self.nested.items():
+            (
+                nested_additions,
+                nested_deletions,
+                nested_modifications,
+            ) = nested_diff.get_diff_display_entries(indent + 2)
+
+            # If there are no changes, skip this nested diff
+            if (
+                len(nested_additions) == 0
+                and len(nested_deletions) == 0
+                and len(nested_modifications) == 0
+            ):
+                continue
+            elif len(nested_deletions) == 0 and len(nested_modifications) == 0:
+                # If there are only additions, display the nested entry as an addition
+                my_additions += [
+                    click.style(f"{' ' * indent}+ {key}:", fg="green")
+                ] + nested_additions
+            elif len(nested_additions) == 0 and len(nested_modifications) == 0:
+                # If there are only deletions, display the nested entry as a deletion
+                my_deletions += [
+                    click.style(f"{' ' * indent}- {key}:", fg="red")
+                ] + nested_deletions
+            else:
+                # Otherwise, display the nested entry as a modification
+                my_modifications += (
+                    [click.style(f"{' ' * indent}~ {key}:", fg="yellow")]
+                    + nested_additions
+                    + nested_deletions
+                    + nested_modifications
+                )
+
+        return (my_additions, my_deletions, my_modifications)
+
+
+# Union type representing the status of a managed element - can either
+# return the (potentially empty) diff between the configured and deployed
+# stack, or an error.
+ManagedElementCheckResult = Union[ManagedElementDiff, ManagedElementError]
+
+
+class ManagedElementReconciler(ABC):
+    """
+    Base class which defines the interface for checking and reconciling a managed element.
+
+    Typically, the constructor will take in a set of resources or user configuration. The
+    implementations of the check and apply methods will then use this configuration to
+    determine the diff between the configured and deployed stack, and to apply the
+    configuration to the deployed stack.
+    """
+
+    @abstractmethod
+    def check(self) -> ManagedElementCheckResult:
+        """
+        Returns whether the user provided config for the managed element is in sync with the external resource.
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def apply(self) -> ManagedElementCheckResult:
+        """
+        Reconciles the managed element with the external resource, returning the applied diff.
+        """
+        raise NotImplementedError()
```

Comment:
I think it would be good to cross reference whatever other reconciliation based systems you would compare this approach to in like a talk about it (terraform, react, k8s, etc) and include in the summary why you choose the language you did. I am not sure how all those systems handle their reconciliation APIs but if theres something common or a particular system this feels closest to, might be good to align against that language.

Reply:
I've added a bit more explanation to the PR summary above, and settled on `plan` and `apply` as top-level action verbs (each of which returns a `diff`).

Reply:
I've added a bit of a write-up above as to why I settled on the naming that's in the PR at present.

I don't have very strong feelings, so would be open to changing `check` towards `diff` or `plan`. 

My slight preference towards `check` is that it expresses the intent behind the action more clearly than `plan`, and to avoid overloading with the noun-form `diff`. This operation is less standard between other state reconciliation tools, so it feels ok to me.